app/certificates.go

```diff
@@ -0,0 +1,441 @@
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return caPEMBuffer.Bytes(), caPrivateKeyPEMBuffer.Bytes(), nil
+}
+
+type generateCertificateInput struct {
+	Organization     string `validate:"required"`
+	OrganizationUnit string
+
+	Duration        time.Duration `validate:"required"`
+	CaPem           []byte        `validate:"required"`
+	CaPrivateKeyPEM []byte        `validate:"required"`
+}
+
+func parseCACerts(caPem, caPrivKeyPem []byte) (*x509.Certificate, any, bool, error) {
+
+	pemBlock, _ := pem.Decode(caPem)
+	if pemBlock == nil {
+		return nil, nil, false, fmt.Errorf("decoding ca cert failed")
+	}
+	caCert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("decoding ca cert failed: %s", err)
+	}
+	pemBlock, _ = pem.Decode(caPrivKeyPem)
+	if pemBlock == nil {
+		return nil, nil, false, fmt.Errorf("decoding ca private key failed")
+	}
+	caPrivateKey, err := x509.ParsePKCS8PrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("parsing ca private key failed: %s", err)
+	}
+	_, isRSA := caPrivateKey.(*rsa.PrivateKey)
+	return caCert, caPrivateKey, isRSA, nil
+}
+
+func generateCertificate(
+	input generateCertificateInput,
+) (certPEM, certPrivateKeyPEM []byte, err error) {
+	validator := validator.New()
+	if err := validator.Struct(input); err != nil {
+		return nil, nil, err
+	}
+	caCert, caPrivateKey, isRSA, err := parseCACerts(input.CaPem, input.CaPrivateKeyPEM)
+	if err != nil {
+		return nil, nil, err
+	}
+	randomLetters, err := generateRandomString(4)
+	if err != nil {
+		return nil, nil, err
+	}
+	subject := pkix.Name{
+		Organization:       []string{input.Organization},
+		OrganizationalUnit: []string{input.OrganizationUnit},
+	}
+	dnsRoot := fmt.Sprintf("client.endentity.%s.%s", input.Organization, randomLetters)
+	now := time.Now()
+	conf := &x509.Certificate{
+		SerialNumber:          big.NewInt(2019),
+		Subject:               subject,
+		NotBefore:             now.Add(-time.Minute), // grace of 1 min
+		NotAfter:              now.Add(input.Duration),
+		BasicConstraintsValid: true,
+		DNSNames:              []string{dnsRoot},
+	}
+	var privateKey any
+	if isRSA {
+		privateKey, err = rsa.GenerateKey(rand.Reader, 4096)
+	} else {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	}
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to generate private key: %v", err)
+	}
+
+	var publicKey any
+	switch k := privateKey.(type) {
+	case *rsa.PrivateKey:
+		publicKey = &k.PublicKey
+	case *ecdsa.PrivateKey:
+		publicKey = &k.PublicKey
+	}
+	cert, err := x509.CreateCertificate(rand.Reader, conf, caCert, publicKey, caPrivateKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEMBuffer := new(bytes.Buffer)
+	err = pem.Encode(certPEMBuffer, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	privBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
```

Comment:
note: iirc Java SDK requires PKCS12 cert format....can you confirm with SDK folks if that is the case? If so, might need a flag...

Reply:
tested with latest Java SDK client, PKCS8 seems to work